### DISCUSSION TITLE
The corret/official way with dict() when taking care of datetime or timedelta.

### DISCUSSION THREAD

**Discussion by xmsley614** [pydantic/pydantic#5558] [Question] [answered]
When taking care of filed like task_duration or task_start_time, if always make me confused.

```python
from datetime import timedelta, datetime

from pydantic import BaseModel

class Foo(BaseModel):
    task_duration: timedelta
    task_start_time: datetime

foo = Foo(task_duration=3660, task_start_time=datetime.now())

print(foo.dict())
```

The out put is like that:
`{'task_duration': datetime.timedelta(seconds=3660), 'task_start_time': datetime.datetime(2023, 4, 23, 17, 25, 43, 302755)}`

What I want is:
`{'task_duration': '1h1m', 'task_start_time': '2023-04-23 17:25`}`

I tried some ways to solve this.
```python
class Foo(BaseModel):
    task_duration: timedelta
    task_duration_str: Optional[str]
    task_start_time: datetime
    task_start_time_str: Optional[str]
    
    def dict(
        self,
        *,
        include: Optional[Union['AbstractSetIntStr', 'MappingIntStrAny']] = None,
        exclude: Optional[Union['AbstractSetIntStr', 'MappingIntStrAny']] = None,
        by_alias: bool = False,
        skip_defaults: Optional[bool] = None,
        exclude_unset: bool = False,
        exclude_defaults: bool = False,
        exclude_none: bool = False,
    ) -> 'DictStrAny':
        self.task_duration_str = convert_functions_a(self.task_duration)
        self.task_start_time_str = convert_functions_b(self.task_start_time)
        return super().dict(
            include=include,
            exclude=exclude,
            by_alias=by_alias,
            skip_defaults=skip_defaults,
            exclude_unset=exclude_unset,
            exclude_defaults=exclude_defaults,
            exclude_none=exclude_none
        )
```

But this will output all fields only if you set the exclude params.
I also tried to use third party libs like [https://github.com/Maydmor/pydantic-computed](https://github.com/Maydmor/pydantic-computed), but I am not quite satisfied.

Then I tried create sub class of datetime and time delta, and override __repr__() functions. But the output will be the same when using dict().
```python
class CustTimeDelta(timedelta):
    def __repr__(self):
        return convert_functions_a(self)
```

I tried to search "pydantic datetime format dict()" and some other key words online, and still can not find out a good solutions.
Is there any solutions? Or what magic method will BaseModel.dict() function call? So I can override it.

**Comment by samuelcolvin** [comment, marked as answer]
You could implement your own type to store both the original value and the validated value, then use `json_encoder` to customize output to JSON, in v1. 

You could do this much more comprehensively with pydantic v2.

Unfortunately I'm v1 it's hard to solve this problem completely.

**Reply by xmsley614** [reply]
`json_encoder `is a much better way. 
I will try it.
Thanks.
